File: 1020 Partition Array Into Three Parts With Equal Sum.java
class Solution {
    public boolean canThreePartsEqualSum(int[] A) {
        int sum = 0;
        for (int num: A) sum += num;
        if (sum % 3 != 0) return false;
        int part = sum / 3;
        int i = 0, s = 0;
        for (i = 0; i < A.length - 1; ++i) {
            s += A[i];
            if (s == part) break;
        }
        if (s != part) return false;
        s = 0;
        int j = 0;
        for (j = A.length - 1; j > i; --j) {
            s += A[j];
            if (s == part) break;
        }
        
        return s == part;
    }
}

/*
// Better solution
class Solution {
    public boolean canThreePartsEqualSum(int[] A) {
        int sum = 0;
        for (int num : A) sum += num;
        if (sum % 3 != 0) return false;
        int part = sum / 3;
        int s = 0, cnt = 0;
        for (int num : A) {
            s += num;
            if (s == part) {
                cnt += 1;
                if (cnt == 2) return true;
                s = 0;
            }            
        }
        return false;
    }
}
*/

File: 1020 Partition Array Into Three Parts With Equal Sum.py
class Solution:
    def canThreePartsEqualSum(self, A: List[int]) -> bool:
        if len(A) < 3: return False
        
        total = sum(A)
        if total % 3 != 0: return False
        part = total // 3
        
        i, s = 0, 0        
        for i in range(len(A)-1):
            s += A[i]
            if s == part: break
            
        if s != part: return False
        
        j, s = 0, 0
        for j in range(len(A)-1, i, -1):
            s += A[j]
            if s == part: break
        
        return s == part

'''
# Beat 100%
class Solution:
    def canThreePartsEqualSum(self, A: List[int]) -> bool:
        total = sum(A)
        if total % 3 != 0: return False
        cnt = s = 0
        for num in A:
            s += num
            if s == part:
                cnt += 1
                if cnt == 2:
                    return True
                s = 0
        return False

'''